File: app/src/main/java/com/adedom/covid19_th/MainViewModel.java
package com.adedom.covid19_th;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class MainViewModel extends ViewModel {

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    private DefaultCovid19Repository repository = new DefaultCovid19Repository();

    private MutableLiveData<Covid19Response> _covid19 = new MutableLiveData<>();
    private MutableLiveData<Throwable> _error = new MutableLiveData<>();

    LiveData<Covid19Response> covid19() {
        return _covid19;
    }

    LiveData<Throwable> error() {
        return _error;
    }

    void fetchCovid19() {
        repository.fetchCovid19()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new SingleObserver<Covid19Response>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        compositeDisposable.add(d);
                    }

                    @Override
                    public void onSuccess(Covid19Response covid19Response) {
                        _covid19.setValue(covid19Response);
                    }

                    @Override
                    public void onError(Throwable e) {
                        _error.setValue(e);
                    }
                });
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.clear();
    }

}

File: app/src/main/java/com/adedom/covid19_th/MainViewModel.kt
package com.adedom.covid19_th

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.*
import kotlin.coroutines.CoroutineContext

class MainViewModel : ViewModel(), CoroutineScope {

    private val job = SupervisorJob()
    private val exceptionHandler = CoroutineExceptionHandler { _, err ->
        _error.value = err
    }
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main + exceptionHandler

    private val repository = DefaultCovid19Repository()

    private val _covid19 = MutableLiveData<Covid19Response>()
    val covid19: LiveData<Covid19Response>
        get() = _covid19

    private val _error = MutableLiveData<Throwable>()
    val error: LiveData<Throwable>
        get() = _error

    fun fetchCovid19() {
        launch {
            try {
                val response = repository.fetchCovid19()
                _covid19.value = response
            } catch (e: Throwable) {
                _error.value = e
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        coroutineContext.cancel()
    }

}
